fenix/app/src/main/java/org/mozilla/fenix/HomeActivity.kt

```diff
@@ -868,14 +875,18 @@ open class HomeActivity : LocaleAwareAppCompatActivity(), NavHostActivity {
      * private mode directly before the content view is created. Returns the mode set by the intent
      * otherwise falls back to the last known mode.
      */
-    internal fun getModeFromIntentOrLastKnown(intent: Intent?): BrowsingMode {
+    internal fun setModeFromIntent(intent: Intent?) {
         intent?.toSafeIntent()?.let {
             if (it.hasExtra(PRIVATE_BROWSING_MODE)) {
                 val startPrivateMode = it.getBooleanExtra(PRIVATE_BROWSING_MODE, false)
-                return BrowsingMode.fromBoolean(isPrivate = startPrivateMode)
+                // Since the mode is initially set from settings during AppAction.Init, this action
+                // will overwrite the state once the action is processed in the queue if called from
+                // onCreate.
+                components.appStore.dispatch(
+                    AppAction.ModeChange(BrowsingMode.fromBoolean(isPrivate = startPrivateMode)),
+                )
```

Comment:
There is a bit of a race condition with the UI here, but I don't anticipate that it will result in any (more) perceivable jank. Plus, now we aren't blocking the main thread while we read from settings so this may end up being an improvement in reactivity